### ISSUE TITLE
Custom js

### ISSUE THREAD

**Issue by choldgraf** [jupyter-book/myst-theme#437]

**Comment by nthiery** [comment]
And in JupyterBook: https://jupyterbook.org/en/stable/advanced/html.html#custom-css-or-javascript

**Comment by agoose77** [comment]
@rowanc1 my gut feeling is that we want to discourage people from embedding web markup / scripting inside singular pages. I wonder whether our UX can be built around global files / styles that are gated with the proper tags. We can discuss when we refine this work for beta.

**Comment by rowanc1** [comment]
Thanks @cboettig and @nthiery for your help here. Do you have examples to point to online/repo which require these scripts to be loaded? And/or use cases (beyond the leafmap/altair etc. that you mentioned) that can help inform our thinking around this?

**Comment by cboettig** [comment]
@rowanc1 e.g. https://espm-157.carlboettiger.info/spatial-2, rendered from https://github.com/espm-157/website/blob/main/reading/spatial-2.ipynb.  note the desperate use of iframe :-)

**Comment by nthiery** [comment]
Here is my most immediate use case: for my online course notes, I wrote a tiny
JS+CSS library that lets the read show/hide solutions and notes for the instructors.
Here is an example with jupyter-book (sorry, in French); you can click on the solutions
to unfold them, or user the two sliders at the top of the page:

https://nicolas.thiery.name/Enseignement/Info111/Semaine2/TD.html

I am in the process of migrating to mystmd and would like to keep the same feature.

(for the curious: the solutions are marked in the notebook/md sources using nbgrader's
syntax; notes for instructors are written as admonitions with a special title).

That use case illustrates well a nice feature of  jupyter-book / sphinx: that you can easily
customize the output to tackle special needs.

**Comment by choldgraf** [comment]
Two more use cases: both using an external library to enrich page content

### A directive for social media previews

In my personal blog, I have a Sphinx extension called `socialpost` that lets you link to social media posts like this:

 ````
 ```{socialpost} https://x.com/ProjectJupyter/status/1850964152082448459
 ```
 ````

 It generates a block of HTML that Twitter specifies for this kind of thing:

```
<blockquote class="twitter-tweet"><p lang="en" dir="ltr">
  <a href="https://twitter.com/choldgraf/status/1564614538309390345">Tweet from @choldgraf</a>
</blockquote>
```

 And on page load, a JS library renders it into an interactive and nicely-formatted social preview

### A library to enrich table output

2i2c maintains a [list of running hubs](https://infrastructure.2i2c.org/reference/hubs/#table-of-all-running-hubs) for all the communities we're currently working with. This is stored as a "regular table" and I believe it is loaded with the CSV-table list.

On that page, we load the [datatables library](https://datatables.net/) and then use the configuration below to load it and point it to our "normal" hubs table to turn it into a nice interactive one:

https://github.com/2i2c-org/infrastructure/blob/6c86b11ede01c25b33923a31bff3376ebdaaa0df/docs/reference/hubs.md?plain=1#L39-L63

**Comment by stevejpurves** [comment]
@cboettig I take it you mean that ipyleaflet, ipywidgets do not appear without a kernel connection from the myst site yes?
If so for those cases, getting https://github.com/jupyter-book/myst-theme/pull/429 over the line should handle those.

For altair, we expect those plots to appear already without further work.

In other cases, where jupyter outputs like bokeh, etc... that require a lab extension, that would be better addressed by mystmd shipping the extension code and loading it (as opposed to end user intervention to get the scripts), we can look towards a jupyterlab-esque loading mechanism that can automatically get the correct version of extension JS code at site build time.

For some of the other cases mentioned, perhaps new directives should be considered?

**Comment by agoose77** [comment]
As @stevejpurves points out, there are some features for which we should really be incorporating features into our themes, e.g. widget/renderer support. For one-off widgets, one feature that we discussed at SciPy 2024 was the inclusion of `anywidget` as a mechanism for front-end interactivity. This would allow users to author trivial "widgets" that render DOM nodes, if they want custom extensions. 

Like Jupyter Notebok 6 -> Jupyter Notebook 7 / Lab, a general theme for MyST vs JB (in my view) is to treat web exports more formally (by building a rich document model) so that all export types can leverage new features, and so that our documents aren't as fragile as e.g. Jupyter Book. For example, a core benefit of MyST document ASTs is that you can render them with _new_ frontends after build, meaning that you can e.g. distribute content building but centralise the view layer. As such, the document-view separation is something we want to do as much work to preserve as possible.

For use-cases like @nthiery's solution show/hide, that should be ideally something that the theme takes charge of. Perhaps we can start to collect those kinds of usages together, and identify some useful theme components.

**Comment by cboettig** [comment]
Thanks @stevejpurves, sorry I was unclear.  Yes, those maps are pure javascript (maplibre-js, not leaflet-js in this case); but if you were to run the same python code in a Jupyter notebook, I believe `leafmap` renders that map using [maplibre's ipywidget](https://eoda-dev.github.io/py-maplibregl/jupyter/)?  Apologies I'm not really up to speed on the internals, but I'll keep an eye on #429, that sounds fantastic, thanks!

**Comment by nthiery** [comment]
> For use-cases like [@nthiery](https://github.com/nthiery)'s solution show/hide, that should be ideally something that the theme takes charge of.

Yes, it would be great indeed to have a theme for teaching purposes! Nonetheless, I believe it's nice to provide users with a quick js/css extension point to play around with ideas until they mature, like what JB provides.

**Comment by bryanwweber** [comment]
Another use case I have is to embed threejs figures or SVG + JS into the site, see here: https://orbital-mechanics.space/the-n-body-problem/two-body-inertial-numerical-solution.html#fig-two-body-inertial which is done by something like (in JB1):

````
```{code-cell} python
from myst_nb import glue
from IPython.display import HTML

glue("html_js_figure", HTML("path to html + js file"), display=False)
```
Text text text text
:::{glue:figure} html_js_figure
:::
````

In JB2, it'd be nice if the cell output could still be used in this same way, something like:

````
```{code-cell} python
:label: html-js-figure

from IPython.display import HTML
HTML("path to html + js file")
```
Text text text
:::{figure} #html-js-figure

Caption
:::
````
similar to the way Matplotlib or Altair plots are shown in the docs.

**Comment by ryanlovett** [comment]
My institution is requiring the addition of an analytics <script> into <head> or <body> on every page. For my purposes, it would be sufficient to have `site.options.scripts` where an author could specify something like:

```yaml
site:
  options:
    scripts:
      - src: https://example.com/script.js
        attributes:
          async: true
```

I think this could be enabled in `themes/book/app/root.tsx` but I need to test.

Is this approach worth pursuing?

**Comment by choldgraf** [comment]
@ryanlovett personally I think that pattern is reasonable to me. I wonder if there are any objections in-principle to this feature from @agoose77 or @stevejpurves or if this is just a matter of prioritizing and implementing a solution. It feels like low-hanging fruit if it isn't that big of a deal to implement...

**Comment by agoose77** [comment]
@ryanlovett can you clarify whether the analytics is plausible.js, or something else?

**Comment by ryanlovett** [comment]
@agoose77 It is from siteimprove.

https://siteimproveanalytics.com/js/siteanalyze_6294756.js

I'm not sure if it is customized for my institution or not.

**Comment by ryanlovett** [comment]
They describe the scripts at https://help.siteimprove.com/support/solutions/articles/80000448448-adding-siteimprove-analytics-javascript-to-your-website.

So perhaps for my specific issue, it can be added via `site.options.analytics_siteimprove` with a value of the account ID.

**Comment by tobydriscoll** [comment]
I would like to add the copy-tex extension for Katex so that math is copied as latex source. So that's another potential use of this feature.

**Comment by choldgraf** [comment]
This is sort-of supported now, via the anywidget interface:

https://mystmd.org/guide/widgets

it's not the same thing as full-blown javascript on the page but you can do a lot of the same stuff with it
